### PR TITLE
remove url formatting on slack messages

### PR DESCRIPTION
**Pull request by freibuis**
Currently url formatting breaks plugins like whois which require original un-formatted url to work.
<http://www.domain.com|www.domain.com> -> www.domain.com
<http://www.domain.com> -> http://www.domain.com

This only un-formats urls. more work would be needed to unformat every thing else that slacks chat engine changes. if slack changes its formating in the future this will break!

this should make most plugins that use urls work. 

sorry I currently have no l clue how to test this

### CODE REVIEW

**Comment by jimmycuadra** [comment]
Thanks for the PR! The fact that Slack does not provide the unformatted version of a message via the API has been quite frustrating for our use case. While this may get the job done, I think we want to approach it the way Slack did for the Hubot adapter. You can see a lot of important history on this topic by reading through slackhq/hubot-slack#109, slackhq/hubot-slack#129, and slackhq/hubot-slack#131.

**Comment by jimmycuadra** [comment]
tl;dr: Look at [SlackBot#removeFormatting](https://github.com/slackhq/hubot-slack/blob/f900ec8cf9949d3c32b8503ab1b7ab08d36878c6/src/slack.coffee#L153)

**Comment by freibuis** [comment]
its looks like pull request #11 was pretty close to what https://github.com/slackhq/hubot-slack/blob/f900ec8cf9949d3c32b8503ab1b7ab08d36878c6/src/slack.coffee#L153 already had just needed to change regex order to match hubut-slack project

Wrote a test for a few possible formatting issues that I have seen when working with plugins.

I hope these tests are good enough.

let me know what you would like to change

**Comment by jimmycuadra** [comment]
Thanks! We're getting close here. There are a lot of conditions inside the format removal code and we want to verify that they all work as expected, so we should expand the test coverage. Luckily hubot-slack has already taken care of outlining the test cases for us, so they can probably be more or less copied directly: https://github.com/slackhq/hubot-slack/blob/f900ec8cf9949d3c32b8503ab1b7ab08d36878c6/test/slack.coffee#L18 It'd also be useful to split the regex across multiple lines with a common for each important component, like how it's formatted in the hubot-slack code.

A few notes for style differences: in lita-slack, test cases should use be written in the present tense without "should," e.g. "does nothing if there are no user links" instead of "should do nothing if there are no user links" for the first test case in hubot-slack. In the implementation code, we should use `&&`, `||`, and `!` instead of `and`, `or`, and `not`. Parens are also preferred for all method calls even though they are optional in Ruby in most cases.

Thank you for your work on this. Let me know if any more guidance is needed!

**Comment by freibuis** [comment]
@jimmycuadra  no problem. nut shell break up regex and more tests?

**Comment by freibuis** [comment]
I have had to make some slight behavior changes to the output of urls. 

Hubot-Slack project outputs 
```
url <http://www.domain.com> -> http://www.domain.com 
fqdn:<http://www.domain.com|www.domain.com> -> http://www.domain.com
```
I Believe that the second is wrong and will break tools that require domain names. thinking of whois etc

This PR uses the following logic
```
url <http://www.domain.com> -> http://www.domain.com
fqdn <http://www.domain.com|www.domain.com> -> www.domain.com
```

Channel Mapping is included in this PR because its required to parse messages with `<@123|user> ` references

all tests copied from the hubot-slack project.

**Comment by jimmycuadra** [comment]
Sorry I've been so slow to respond to this PR – I really appreciate your work on this. I'm gonna be making some other changes to lita-slack very soon and will get your work integrated then.

**Comment by freibuis** [comment]
No Problem. I hope this is what you where looking for,

Once this has been integrated I will map all the Slack Events that are currently missing.
https://api.slack.com/rtm

I will need to fix events like channel_rename.

**Comment by jimmycuadra** [comment]
Released in lita-slack 1.5! Thanks very, very much for your work on this! It was a long-desired feature, now finally implemented. :D

**Comment by Ginja** [comment]
Just updated the lita-slack adapter to 1.5, and happy to report that `pager identify` now works. Thanks for the info!
